### PR TITLE
Crop on the ID-selected region(s) and not on the whole shapefile

### LINKED ISSUE #1072: extract_shape - crop on the ID-selected region and not on the whole shapefile
**Describe the bug**
PR ESMValGroup/ESMValTool#2120 adds shapefiles for the scientific SREX and AR6 regions to be used by extract_shape in the preprocessor. 

I understand the documentation 
> crop: by default extract_region will be used to crop the data to a minimal rectangular region containing the shape. Set to false to only mask data outside the shape. Data on irregular grids will not be cropped.
(https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/preprocessor.html?highlight=extract_shape#extract-shape)

To say that longitudes and latitudes with no value inside of the given shape will be dropped. However, this does not seem to be the case. 

[recipe_climwip_test_basic.yml.txt](https://github.com/ESMValGroup/ESMValCore/files/6334231/recipe_climwip_test_basic.yml.txt)
[shapefiles.zip](https://github.com/ESMValGroup/ESMValCore/files/6334298/shapefiles.zip)

Example output file:
```bash
ncdump -h recipe_climwip_test_basic_20210419_074811/preproc/calculate_weights_climwip/tas_CLIM/CMIP5_ACCESS1-0_Amon_historical-rcp85_r1i1p1_tas_1995-2014.nc
netcdf CMIP5_ACCESS1-0_Amon_historical-rcp85_r1i1p1_tas_1995-2014 {
dimensions:
	lat = 72 ;
	lon = 144 ;
	bnds = 2 ;
```
Can anyone recreate this?

### PR DESCRIPTION
**Pull request by stefsmeets**
## Description

Hi all! This PR fixes a bug where the the `extract_shape` preprocessor crops on the whole shapefile when a subset of ids is specified. For example, if the whole shapefile covers regions all over the world, let's say A, B and C, but you are only interested in A. `extract_region` currently will return a cube with coordinates defined by a bounding box fitting A, B and C, masking everything 
outside of A. This can result in unexpectedly large cubes.

This PR crops on the selected regions, so the bounding box will fit _just_ A. This works for multiple regions, e.g. if A and B are specified, the bounding box will fit A and B (but not C).

This only applies if the `ids` keyword is specified. If `ids` is not specified (`None`), there is no change in behaviour.

Closes #1072 

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Review by lukasbrunner — APPROVED** [review]
Thanks for implementing this Stef! The code looks good to me and runs for me. It also solves the issue for the recipe where it first appeared for me (https://github.com/ESMValGroup/ESMValTool/blob/add_recipe_climwip/esmvaltool/recipes/recipe_climwip_brunner2019_med.yml). Note, however, that I had to comment out the land-sea masking from the recipe due to #1096

**Comment by stefsmeets** [comment]
Hi @ESMValGroup/esmvaltool-coreteam This PR has been reviewed and is ready to be merged 👍

**Comment by valeriupredoi** [comment]
nice one @stefsmeets :beer: Took a cursory look too, but pinging @jvegasbsc for a serious, Iberian technical review :grin:

File: esmvalcore/preprocessor/_area.py [hunk 1]
```diff
@@ -520,8 +573,15 @@ def extract_shape(cube, shapefile, method='contains', crop=True,
             pad_hawaii = True
 
         if crop:
+            lon_min, lat_min, lon_max, lat_max = _get_bounds(
```
Comment (zklaus) [inline esmvalcore/preprocessor/_area.py:576]: lon and lat seem to flipped compared with line 457. Which is correct? The other one should be changed. And if possible let's make sure that the tests would catch this, i.e. choose a region for tests with longitudes outside [-90., 90.].

**Review by zklaus — CHANGES_REQUESTED** [review]

File: esmvalcore/preprocessor/_area.py (hunk 1 above)
Comment (stefsmeets) [reply esmvalcore/preprocessor/_area.py:576]: Good find, I will have a look on Friday. I always mix those around 😅

Comment (zklaus) [reply esmvalcore/preprocessor/_area.py:576]: Alright, I'll bump it to 2.4.0 then.

Comment (stefsmeets) [reply esmvalcore/preprocessor/_area.py:576]: Should be fixed now 👍

The code was working correctly, I just mixed up the variable names in the function. The tests would catch this if the coordinates were be flipped (I just checked).

**Review by zklaus — APPROVED** [review]

**Pull request merged**